Adding a "Test-Driven Development" document (#579)
exercism/docs: using/solving-exercises/tdd.md

# Test-Driven Development (TDD)

Test-Driven Development (sometimes called Test-First Development or Test-Driven Design) is the practice of writing the unit tests first, before you write a single line of implementation code.

## On Exercism, the tests _are_ the requirements!

> Each exercise you work on will have some instructions describing in general terms what you need to do.

Reviewer: ```suggestion
All Practice Exercises you work on (those ones that don't teach you a new concept) will have some instructions describing in general terms what you need to do.
```

By design, these instructions do not account for programming-language-specific implementation details because they are shared by all of Exercism’s 70+ language tracks.
Some language tracks will append more specific details for you, but not all of them do.

> When you start working on an exercise, give the instructions a careful read.

Reviewer: ```suggestion
When you start working on a Practice Exercise, give the instructions a careful read.
```

They will give you a broad overview of how you go about implementing a solution.
But you will have to read the _tests_ to understand the full and exact requirements:

- Must the result be a particular kind of data structure?
- Must the result be sorted in some order?
- How are you expected to handle exceptions? And so on.

You have solved an exercise when all the provided tests run and pass.

> In other words, your solution is not just an interpretation the instructions that "looks right", your solution is a program that _satisfies the given tests_.

Reviewer 2: ```suggestion
In other words, your solution is not just an interpretation of the instructions that "looks right", your solution is a program that _satisfies the given tests_.
```

**The tests represent the complete requirements for the exercise.**

> ## Rationale for Test-Driven Development

Reviewer: I'd be tempted to move this even further, below "## Working locally" - just because I feel for 95% of people, the details they're reading this for are the skips and the "what do I do". The more I work with people the more I realise that the "why?" comes a long way after the "what is the immediate thing I need to do to satisfy myself" in their priorities, and if its not very quickly obvious, they don't read.

> While it may seem like putting the cart before the horse, there are several good reasons why you might want to write unit tests before writing the implementation code:

Reviewer: ```suggestion
While it may seem like "putting the cart before the horse", there are several good reasons why you might want to write unit tests before writing the implementation code:
```

Just as that might be weird for non-native readers without quotes.

Test results will default to expanding the first failed test message (although what is displayed varies by track).

Don’t be discouraged by a large number of failing tests.
Focus on making them pass one-by-one.

## Working locally

Many tracks use "skipped" tests in their test files.
Initially, only the first test is "active" and the remaining are deactivated (how this happens varies by track).
When you run the test suite in your environment, only the first test runs.
We do this to encourage you to follow this workflow:

1. Before adding any new code, run the test suite: you should see a failing test.
1. Add _just enough_ code to pass the test.
1. Run the test suite. If the test still fails, repeat the last step.
1. Review your code and refactor as desired, making sure the tests still pass.
1. Once you have passed all the tests, congratulations you're done!
1. Otherwise, "unskip" the next test and goto 1. 

Exactly how tests are "unskipped" (or selected) depends on the track.
For some tracks, it might be commenting or removing an annotation.
For some tracks, it might be changing an attribute from true to false.

> Take the time to read the track documentation: it will explain these details.

Reviewer: Maybe a link to https://exercism.org/docs/tracks here?

By design, these instructions do not account for programming-language-specific implementation details because they are shared by all of Exercism’s 70+ language tracks.
Some language tracks will append more specific details for you, but not all of them do.

When you start working on a Practice Exercise, give the instructions a careful read.
They will give you a broad overview of how you go about implementing a solution.
But you will have to read the _tests_ to understand the full and exact requirements:

- Must the result be a particular kind of data structure?
- Must the result be sorted in some order?
- How are you expected to handle exceptions? And so on.

You have solved an exercise when all the provided tests run and pass.
In other words, your solution is not just an interpretation of the instructions that "looks right", your solution is a program that _satisfies the given tests_.
**The tests represent the complete requirements for the exercise.**

## How does Exercism apply TDD?

We've done the work of writing a unit test suite for you.

> Your goal is to write a solution that contains just enough code to make all those unit tests pass, and no more.

Reviewer 2: Phrasing it that way might discourage users from writing a more complete implementation. What about changing it to:
```suggestion
Your goal is to write a solution that contains just enough code to make all those unit tests pass.
You can extend your solution further if you'd like.
```

Author: I dunno. That phrasing goes against TDD. Maybe be explicit about it:

> Your goal is to write a solution that contains just enough code to make all those unit tests pass, and no more.
> You are welcome to extend your solution beyond that, but that's not TDD anymore.

And then we could add

> Mentoring sessions often venture into the territory beyond the given tests.

Reviewer 2: True, but the current phrasing suggests we are only focused on TDD and discourages anything beyond it. This approach kind of defeats Exercism's purpose as a learning platform. Considering that few users request mentoring, I'm unsure mentoring sessions alone could address this. Don't we want to encourage users to create more complete and thoughtful implementations? It might be worth adding a section explaining what users can do after passing the tests.

Author: good point

# Test-Driven Development (TDD)

Test-Driven Development (sometimes called Test-First Development or Test-Driven Design) is the practice of writing the unit tests first, before you write a single line of implementation code.

## On Exercism, the tests _are_ the requirements!

All Practice Exercises you work on (those ones that don't teach you a new concept) will have some instructions describing in general terms what you need to do.

> By design, these instructions do not account for programming-language-specific implementation details because they are shared by all of Exercism’s 70+ language tracks.

Reviewer 3: This line idea a smart apostrophe. Line 25 uses a single quote. It would be good to be consistent with the contractions.

Reviewer 4: Line 7 as well, so even "close" example of inconsistency.

Should you choose to work with a mentor (and we encourage you to do that once you get the tests passing), they can help you refactor and refine your initial implementation, or even propose new unit tests.

## Working in the online editor

When you’re working in the code editor on Exercism’s website, you can read the tests but you are not able to edit them.
All tests will be executed each time you run them, regardless of any "skip" mechanisms noted in the test file.

Test results will default to expanding the first failed test message (although what is displayed varies by track).

Don’t be discouraged by a large number of failing tests.
Focus on making them pass one-by-one.

## Working locally

Many tracks use "skipped" tests in their test files.
Initially, only the first test is "active" and the remaining are deactivated (how this happens varies by track).
When you run the test suite in your environment, only the first test runs.
We do this to encourage you to follow this workflow:

1. Before adding any new code, run the test suite: you should see a failing test.
1. Add _just enough_ code to pass the test.

> 1. Run the test suite. If the test still fails, repeat the last step.

Reviewer 3: One sentence per line.

1. Review your code and refactor as desired, making sure the tests still pass.
1. Once you have passed all the tests, congratulations you're done!
1. Otherwise, "unskip" the next test and goto 1. 

> Exactly how tests are "unskipped" (or selected) depends on the track.

Reviewer 4: ```suggestion
Exactly how tests are "unskipped" (or activated) depends on the track.
```

For some tracks, it might be commenting or removing an annotation.
For some tracks, it might be changing an attribute from true to false.
Take the time to read [the documentation for your track][track-docs]: it will explain these details.

## Rationale for Test-Driven Development

While it may seem like "putting the cart before the horse", there are several good reasons why you might want to write unit tests before writing the implementation code:

> 1. Design.
>    It forces you to think first about the design of the **interface** to the code, instead of jumping straight to the implementation.
>    Having a well-designed (and testable!) interface is often more important than having an efficient implementation.
> 
> 1. Discipline.
>    Writing tests is often seen as a chore or an afterthought; writing the tests _first_ guarantees that at the end of the day you will have written enough unit tests to cover most or all of your code's functionality (rather than possibly never getting around to it).
> 
> 1. Less Work.
>    If you apply a tight cycle of write one test, then write the code to implement that test, then write the next test, your code ends up growing organically.
>    This often (though not always) leads to less wasted effort; you end up writing all the code you need, and none of the code you don't need.

Reviewer 5: ```suggestion
1. Design.
   It forces you to think first about the design of the **interface** to the code, instead of jumping straight to the implementation.
   Having a well-designed (and testable!) interface is often more important than having an efficient implementation.

2. Discipline.
   Writing tests is often seen as a chore or an afterthought; writing the tests _first_ guarantees that at the end of the day you will have written enough unit tests to cover most or all of your code's functionality (rather than possibly never getting around to it).

3. Less Work.
   If you apply a tight cycle of write one test, then write the code to implement that test, then write the next test, your code ends up growing organically.
   This often (though not always) leads to less wasted effort; you end up writing all the code you need, and none of the code you don't need.
```

Author: This is just changing the numbers? Markdown does that for us.

https://github.com/exercism/docs/blob/c54146127eb7b53bfd42c4af2fdcdd0a48b419ed/using/solving-exercises/tdd.md#rationale-for-test-driven-development

Reviewer 5: TIL!  I didn't know that.  😄   Carry on and reject, then.  Apologies.

Reviewer 4: This matters more when it is expected that it (as a markdown/plain text file) will be read in the terminal, but for rending, definitely  nice that it "auto-adjusts".

1. Before adding any new code, run the test suite: you should see a failing test.
1. Add _just enough_ code to pass the test.
1. Run the test suite.
1. If the test still fails, repeat step 2.
1. Once the test passes, refactor your code as desired, ensuring all the activated tests still pass.
   Refactoring might include:
    - removing any duplicated code,
    - spliting long functions into smaller ones
    - adding comments, etc.
1. "Unskip" the next test and repeat from step 1.

Repeat these steps until you have unskipped all the tests.
Once all the tests are passing, congratulations, you have solved the exercise!

Exactly how tests are "unskipped" (or activated) depends on the track.
For some tracks, it might be commenting or removing an annotation.
For some tracks, it might be changing an attribute from true to false.
Take the time to read [the documentation for your track][track-docs]: it will explain these details.

## Rationale for Test-Driven Development

While it may seem like "putting the cart before the horse", there are several good reasons why you might want to write unit tests before writing the implementation code:

1. Design.

>    It forces you to think first about the design of the **interface** to the code, instead of jumping straight to the implementation.

Reviewer 2: I think many people might find the wording here confusing. It might help to clarify how the tests are connected to the design of the interface.

Author: That text comes straight from https://web.archive.org/web/20220918221108/http://testfirst.org/about#tdd

I'm not particularly inclined to change it, but you can change my mind with a concrete suggestion :)

Reviewer 2: I would have provided a suggestion, but I found the phrase "interface to the code" slightly awkward, therefore I wasn't sure what it was referring to. Assuming the phrase is perfectly fine in this context, I would propose a minor update like so:

```It forces you to think first about the design of the interface to the code (how the code will be used), instead of jumping straight to the implementation```

Author: How about

`It forces you to think first about your program's [interface][api] (how it exposes its functionality to the world), instead of jumping straight into how you will implement the code.`

`[api]: https://en.wikipedia.org/wiki/API`

Reviewer 2: Sure! That's even better :)

You have solved an exercise when all the provided tests run and pass.
In other words, your solution is not just an interpretation of the instructions that "looks right", your solution is a program that _satisfies the given tests_.
**The tests represent the complete requirements for the exercise.**

## How does Exercism apply TDD?

We've done the work of writing a unit test suite for you.
Your goal is to write a solution that contains just enough code to make all those unit tests pass.

Keep this in mind: the TDD approach will help you get to the solution, but you don't need to stop there.
If you want to extend your solution beyond the requirements, you are welcome to do so.
Should you choose to work with a mentor (and we encourage you to do that once you get the tests passing), they can help you refactor and refine your initial implementation, or even propose new unit tests.

## Working in the online editor

When you're working in the code editor on Exercism's website, you can read the tests but you are not able to edit them.
All tests will be executed each time you run them, regardless of any "skip" mechanisms noted in the test file.

> Test results will default to expanding the first failed test message (although what is displayed varies by track).

Reviewer 3: ```suggestion
When there are multiple tests that fail, three website initially only displays the results of the first failure.
You can click on other failures to expand them, too!
Sometimes the first result may not be the most informative.
```

Don't be discouraged by a large number of failing tests.
Focus on making them pass one-by-one.

## Working locally

Many tracks use "skipped" tests in their test files.

> Initially, only the first test is "active" and the remaining are deactivated (how this happens varies by track).

Reviewer 3: ```suggestion
Initially, only the first test is "active" and the remaining are inactive (how this happens varies by track).
```

When you run the test suite in your environment, only the first test runs.
We do this to encourage you to follow this workflow:

1. Before adding any new code, run the test suite: you should see a failing test.
1. Add _just enough_ code to pass the test.
1. Run the test suite.
1. If the test still fails, repeat step 2.

> 1. Once the test passes, refactor your code as desired, ensuring all the activated tests still pass.

Reviewer 3: ```suggestion
1. Once the test passes, refactor your code as desired, ensuring all the activate tests still pass.
```

Reviewer 2: @IsaacG Something's up with your suggestion here. Do you mean:
```suggestion
1. Once the test passes, refactor your code as desired, ensuring all active tests still pass.
```

Reviewer 3: My intention was to change "activated" to "active".

   Refactoring might include:
    - removing any duplicated code,
    - spliting long functions into smaller ones
    - adding comments, etc.
1. "Unskip" the next test and repeat from step 1.

Repeat these steps until you have unskipped all the tests.
Once all the tests are passing, congratulations, you have solved the exercise!

Exactly how tests are "unskipped" (or activated) depends on the track.
For some tracks, it might be commenting or removing an annotation.
For some tracks, it might be changing an attribute from true to false.

> Take the time to read [the documentation for your track][track-docs]: it will explain these details.

Reviewer 3: ```suggestion
Take the time to read [the documentation for your track][track-docs]; it will explain these details.
```

For tracks that don't skip the tests, applying this workflow may be as straightforward as commenting out the tests and uncommenting them one-by-one.

## Rationale for Test-Driven Development

> While it may seem like "putting the cart before the horse", there are several good reasons why you might want to write unit tests before writing the implementation code:

Reviewer 3: ```suggestion
While it may seem like "putting the cart before the horse", there are several good reasons why you might want to write unit tests before writing the implementation code.
```

You have solved an exercise when all the provided tests run and pass.
In other words, your solution is not just an interpretation of the instructions that "looks right", your solution is a program that _satisfies the given tests_.
**The tests represent the complete requirements for the exercise.**

## How does Exercism apply TDD?

We've done the work of writing a unit test suite for you.
Your goal is to write a solution that contains just enough code to make all those unit tests pass.

Keep this in mind: the TDD approach will help you get to the solution, but you don't need to stop there.
If you want to extend your solution beyond the requirements, you are welcome to do so.
Should you choose to work with a mentor (and we encourage you to do that once you get the tests passing), they can help you refactor and refine your initial implementation, or even propose new unit tests.

## Working in the online editor

When you're working in the code editor on Exercism's website, you can read the tests but you are not able to edit them.
All tests will be executed each time you run them, regardless of any "skip" mechanisms noted in the test file.

> When there are multiple tests that fail, three website initially only displays the results of the first failure.

Reviewer 2: ```suggestion
When there are multiple tests that fail, the website initially only displays the results of the first failure.
```

You can click on other failures to expand them, too!
Sometimes the first result may not be the most informative.

Don't be discouraged by a large number of failing tests.
Focus on making them pass one-by-one.

## Working locally

Many tracks use "skipped" tests in their test files.
Initially, only the first test is "active" and the remaining are inactive (how this happens varies by track).
When you run the test suite in your environment, only the first test runs.
We do this to encourage you to follow this workflow:

1. Before adding any new code, run the test suite: you should see a failing test.
1. Add _just enough_ code to pass the test.
1. Run the test suite.
1. If the test still fails, repeat step 2.
1. Once the test passes, refactor your code as desired, ensuring all active tests still pass.
   Refactoring might include:
    - removing any duplicated code,

>     - spliting long functions into smaller ones

Reviewer 2: ```suggestion
    - splitting long functions into smaller ones,
```

Author: Thanks for the sharp eyes!

## Rationale for Test-Driven Development

While it may seem like "putting the cart before the horse", there are several good reasons why you might want to write unit tests before writing the implementation code.

1. Design.
   It forces you to think first about your program's [interface][api] (how it exposes its functionality to the world), instead of jumping straight into how you will implement the code.
   Having a well-designed (and testable!) interface is often more important than having an efficient implementation.

1. Discipline.
   Writing tests is often seen as a chore or an afterthought; writing the tests _first_ guarantees that at the end of the day you will have written enough unit tests to cover most or all of your code's functionality (rather than possibly never getting around to it).

1. Less Work.
   If you apply a tight cycle of write one test, then write the code to implement that test, then write the next test, your code ends up growing organically.
   This often (though not always) leads to less wasted effort; you end up writing all the code you need, and none of the code you don't need.

## Further reading

> * [About Test-First Teaching][test-first] at the archived TestFirst\.org site.
> * [Test-driven development][tdd-wiki] at Wikipedia.
> * [Test Driven Development][tdd-python] on the Python track.

Reviewer 2: This one is for consistency
```suggestion
- [About Test-First Teaching][test-first] at the archived TestFirst\.org site.
- [Test-driven development][tdd-wiki] at Wikipedia.
- [Test Driven Development][tdd-python] on the Python track.
```

1. Before adding any new code, run the test suite: you should see a failing test.
1. Add _just enough_ code to pass the test.
1. Run the test suite.
   If the test still fails, repeat the last step.
1. Review your code and refactor as desired, making sure the tests still pass.
1. Once you have passed all the tests, congratulations you're done!

> 1. Otherwise, "unskip" the next test and goto 1.

Reviewer 2: After carefully reading these steps, I found them a little confusing. Some steps appear to be out of order. May I propose something simpler? (does not let me add suggestion)
```
1. Before adding any new code, run the test suite: you should see a failing test.
1. Add _just enough_ code to pass the test.
1. Run the test suite again.
   If the test still fails, repeat from step 2.
1. When the test passes, unskip the next one and repeat from step 1.
1. Once all tests pass, refactor your code as desired, ensuring tests still pass — congratulations you're done!
```

**Edited**

Author: I want to keep refactoring after each passing test. Pedantically, it's there in TDD.

I'll give it another go, looking at https://en.wikipedia.org/wiki/Test-driven_development#Coding_cycle

Reviewer 2: > I want to keep refactoring after each passing test. Pedantically, it's there in TDD.

That's fine, with the refactoring after each test there isn't much that needs to be modified:
```
1. Before adding any new code, run the test suite: you should see a failing test.
1. Add just enough code to pass the test.
1. Run the test suite again.
   If the test still fails, repeat from step 2.
1. When the test passes, refactor your code as desired, ensuring tests still pass.
1. Unskip the next test and repeat from step 1 until all tests pass.
1. Once all tests pass, congratulations — you're done!
```

Your version would work too with minimal changes.

Reviewer 2: Nice. The steps are more clear now 👍
